### PR TITLE
Fix and better test abstract files and directories support

### PR DESCRIPTION
**Pull request by joshuagl**
**Fixes issue #**: N/A or a continuation of #1009 

**Description of the changes being introduced by the pull request**:
* Improve documentation for `generate_targets_metadata()`
* Add support for abstract storage backends to `generate_timestamp_metadata()`
* Implement a test for abstract storage support – creating and writing the metadata for a repository with an abstract storage backend

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

File: tests/test_repository_tool.py [hunk 1]
```diff
@@ -550,6 +557,168 @@ def test_get_filepaths_in_directory(self):
 
 
 
+  def test_writeall_abstract_storage(self):
```
Comment (woodruffw) [inline tests/test_repository_tool.py:560]: Thanks for adding this!

**Review by woodruffw — APPROVED** [review]
Looks good to me!

File: tests/test_repository_tool.py [hunk 2]
```diff
@@ -550,6 +557,168 @@ def test_get_filepaths_in_directory(self):
...
+      metadata files.
+      """
+
+      from contextlib import contextmanager
+
+
+      @contextmanager
+      def get(self, filepath):
+        file_object = open(filepath + '.tst', 'rb')
+        yield file_object
```
Comment (lukpueh) [inline tests/test_repository_tool.py:581]: 🥇  Finally we have a `yield` in python-tuf. :)

File: tests/test_repository_tool.py [hunk 3]
```diff
@@ -550,6 +557,168 @@ def test_get_filepaths_in_directory(self):
...
+        file_object.close()
+
```
Comment (lukpueh) [inline tests/test_repository_tool.py:583]: Mind adding a blank line for consistency with the rest of `TestStorageBackend`?

File: tests/test_repository_tool.py [hunk 4]
```diff
@@ -550,6 +557,168 @@ def test_get_filepaths_in_directory(self):
...
+  def test_writeall_abstract_storage(self):
+    # Test creation of a TUF repository with a custom storage backend to ensure
+    # that function relying on a storage backend being supplied operate
+    # correctly
+
+
+    class TestStorageBackend(securesystemslib.storage.StorageBackendInterface):
+      """
+      An implementation of securesystemslib.storage.StorageBackendInterface
+      which mutates filenames on put()/get(), such that trying to read the
```
Comment (lukpueh) [inline tests/test_repository_tool.py]: Very nice way of testing this. Would you mind adding one sentence that elaborates on the mutation, i.e. `filename` in memory vs. `filename + ".tst"` on disk?

**Review by lukpueh — APPROVED** [review]
Excellent work, @joshuagl! I have 2 minor requests inline, but this is good to go regardless. 

FYI I ran the new tests excluding your last commit and they successfully catch the then missing AbstractFileStorage support in `generate_timestamp_metadata`. 💯

File: tests/test_repository_tool.py (hunk 3 above)
Comment (joshuagl) [reply tests/test_repository_tool.py:583]: Eagle eyed review. Fixed in a force-pushed update to the patch, thank you.

File: tests/test_repository_tool.py (hunk 2 above)
Comment (joshuagl) [reply tests/test_repository_tool.py:581]: 🙇

File: tests/test_repository_tool.py (hunk 4 above)
Comment (joshuagl) [reply tests/test_repository_tool.py]: Updated to the following in a force-pushed change
> An implementation of securesystemslib.storage.StorageBackendInterface which mutates filenames on put()/get(), translating filename in memory to filename + '.tst' on-disk, such that trying to read the expected/canonical file paths from local storage doesn't find the TUF metadata files.

**Pull request merged**